### ISSUE TITLE
"An error occurred while handling a request"

### ISSUE THREAD

**Issue by amorey** [ory/oathkeeper#373]
**Describe the bug**

Internal error when oathkeeper denies a request

**Reproducing the bug**

1. Follow the configure/build/run instructions on (Configure and Deploy)[https://www.ory.sh/docs/oathkeeper/configure-deploy].

2. Execute the second example under (Authorizing Requests)[https://www.ory.sh/docs/oathkeeper/configure-deploy#authorizing-requests]:

   ```bash
   $ curl -H "Accept: application/json" -X GET http://127.0.0.1:4455/anything/deny
   {"error":{"code":403,"status":"Forbidden","message":"Access credentials are not sufficient to access this resource"}}
   ```

*Server logs*

time="2020-03-06T11:00:53Z" level=error msg="An error occurred while handling a request" code=403 debug= details="map[]" error="Access credentials are not sufficient to access this resource" reason= request-id= status=403 trace="Stack trace: \ngithub.com/ory/oathkeeper/pipeline/authz.(*AuthorizerDeny).Authorize\n\t/home/ory/pipeline/authz/authorizer_deny.go:49\ngithub.com/ory/oathkeeper/proxy.(*RequestHandler).HandleRequest\n\t/home/ory/proxy/request_handler.go:271\ngithub.com/ory/oathkeeper/proxy.(*Proxy).Director\n\t/home/ory/proxy/proxy.go:134\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\t/usr/local/go/src/net/http/httputil/reverseproxy.go:206\ngithub.com/urfave/negroni.Wrap.func1\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:46\ngithub.com/urfave/negroni.HandlerFunc.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29\ngithub.com/urfave/negroni.middleware.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38\ngithub.com/ory/x/reqlog.(*Middleware).ServeHTTP\n\t/go/pkg/mod/github.com/ory/x@v0.0.93/reqlog/middleware.go:140\ngithub.com/urfave/negroni.middleware.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38\ngithub.com/ory/x/metricsx.(*Service).ServeHTTP\n\t/go/pkg/mod/github.com/ory/x@v0.0.93/metricsx/middleware.go:261\ngithub.com/urfave/negroni.middleware.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38\ngithub.com/urfave/negroni.(*Negroni).ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:96\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2802\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1890\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357" writer=JSON
time="2020-03-06T11:00:53Z" level=info msg="completed handling request" measure#oathkeeper-proxy.latency=800566 method=GET remote="172.17.0.1:38134" request=/anything/deny status=403 text_status=Forbidden took="800.566µs"

*Server configuration*

Same configuration described in [(Configure and Deploy)https://www.ory.sh/docs/oathkeeper/configure-deploy#authorizing-requests]

**Expected behavior**

I didn't expect to see a traceback in the logs

**Environment**

* Version: v0.36.0-beta.4+oryOS.17
* Environment: macOS, Docker Desktop 2.2.0.3 (42716), Docker Engine 19.03.5

**Comment by aeneasr** [comment]
I don’t undetstand? The error is a forbidden one which is also expected because the route you are calling is doing that for demonstration purposes.

Also, I do not understand the „traceback“ comment. You can control traces with the log level (LOG_LEVEL=info disables traces).

**Comment by amorey** [comment]
Yes, the `403 Forbidden` response is expected but is the error message and the accompanying trace in the logs expected as well?

**Comment by aeneasr** [comment]
Yes, but maybe we should replace it with a WARN instead of ERR log level. What you think?

**Comment by amorey** [comment]
I think expected log messages should be INFO or DEBUG. ERROR and WARN are usually reserved for more exceptional conditions that indicate bugs or require mitigation:
https://en.wikipedia.org/wiki/Syslog#Severity_level

The text of the message "An error occured while handling a request" doesn't seem expected and I'm not used to seeing tracebacks except in the context of unhandled errors. I couldn't figure out if I was doing something wrong or if there was a problem with Oathkeeper. Did the request trigger a panic or was there an unhandled error in the code? Why did you choose to include the trace in the logging message?

**Comment by aeneasr** [comment]
Because this is needed for audit trails. Access denied errors are errors and can indicate a malicious actor in a system. A WARN is probably more appropriate though.

**Comment by amorey** [comment]
Maybe I'm not understanding something. Is an "Access Denied" response considered an "error" from the point of view of the oathkeeper service?  From my point of view a as a user, a request was successfully made against the `/decisions` endpoint which returned a 403 Forbidden response; end of story.

How is the traceback useful for audit trails? How can I use this for a security audit:
```
Stack trace: 
github.com/ory/oathkeeper/pipeline/authz.(*AuthorizerDeny).Authorize
	/home/ory/pipeline/authz/authorizer_deny.go:49
github.com/ory/oathkeeper/proxy.(*RequestHandler).HandleRequest
	/home/ory/proxy/request_handler.go:271
github.com/ory/oathkeeper/proxy.(*Proxy).Director
	/home/ory/proxy/proxy.go:134
net/http/httputil.(*ReverseProxy).ServeHTTP
	/usr/local/go/src/net/http/httputil/reverseproxy.go:206
github.com/urfave/negroni.Wrap.func1
	/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:46
github.com/urfave/negroni.HandlerFunc.ServeHTTP
	/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29
github.com/urfave/negroni.middleware.ServeHTTP
	/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
github.com/ory/x/reqlog.(*Middleware).ServeHTTP
	/go/pkg/mod/github.com/ory/x@v0.0.93/reqlog/middleware.go:140
github.com/urfave/negroni.middleware.ServeHTTP
	/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
github.com/ory/x/metricsx.(*Service).ServeHTTP
	/go/pkg/mod/github.com/ory/x@v0.0.93/metricsx/middleware.go:261
github.com/urfave/negroni.middleware.ServeHTTP
	/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
github.com/urfave/negroni.(*Negroni).ServeHTTP
	/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:96
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2802
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1890
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357
```

**Comment by aeneasr** [comment]
> Maybe I'm not understanding something. Is an "Access Denied" response considered an "error" from the point of view of the oathkeeper service? From my point of view a as a user, a request was successfully made against the /decisions endpoint which returned a 403 Forbidden response; end of story.

That's just a different viewpoint that I disagree with. For us, an access denied error is an exception as it should not happen on a regular basis, thus it's logged as such. The trace helps debug problems in the Oathkeeper code (as you probably have noticed it's not released on a stable channel yet). Since the trace is in its own field (e.g. when using JSON log output) it's easily filtered out so I don't really get what the problem is.

I hope we can cut the discussion short as the feedback to move the log level to something else is very valuable, but we'll keep the trace for the foreseeable future due to the reasons mentioned.

**Comment by amorey** [comment]
If you want to preserve this information for debugging purposes then I would set the log level to DEBUG and change the language of "An error occured while handling a request" because that implies that an unexpected internal error occured with the oathkeeper service. When I saw the ERROR messages in the log I assumed it was user error and I tried debugging the problem until I gave up.

Another thing you might want to consider is splitting the Oathkeeper log into an "Access" log and an "Error" log like Apache does:
https://httpd.apache.org/docs/2.4/logs.html#accesslog
https://httpd.apache.org/docs/2.4/logs.html#errorlog

**Comment by aeneasr** [comment]
Right, in Hydra we log traces only when `LOG_LEVEL=debug` but apparently this doesn't apply here. We can fix that.

**Comment by aeneasr** [comment]
The fix needs to be applied upstream, in Herodot, in the default error reporter. https://github.com/ory/herodot/blob/master/error_reporter.go#L39 should only be applied when `LOG_LEVEL` >= `debug`

**Comment by amorey** [comment]
While playing around with the kratos quickstart and I noticed that the oathkeeper error stacktrace only appears with JSON requests. For example:

1. Accept: text/html

    ```bash
    $ curl -H 'Accept: text/html' http://127.0.0.1:4455/
    <a href="http://127.0.0.1:4455/auth/login">Found</a>.

    ```

     ```
    oathkeeper_1                  | {"level":"info","method":"GET","msg":"started handling request","remote":"172.18.0.1:53314","request":"/","time":"2020-03-10T18:33:49Z"}
    oathkeeper_1                  | {"error":"Access credentials are invalid","granted":false,"http_host":"127.0.0.1:4455","http_method":"GET","http_url":"http://127.0.0.1:4455/","http_user_agent":"curl/7.64.1","level":"warning","msg":"No authentication handler was responsible for handling the authentication request","reason_id":"authentication_handler_no_match","rule_id":"ory:kratos-selfservice-ui-node:protected","time":"2020-03-10T18:33:49Z"}
    oathkeeper_1                  | {"error":"Access credentials are invalid","granted":false,"http_host":"127.0.0.1:4455","http_method":"GET","http_url":"http://127.0.0.1:4455/","http_user_agent":"curl/7.64.1","level":"warning","msg":"Access request denied","time":"2020-03-10T18:33:49Z"}
    oathkeeper_1                  | {"level":"info","measure#oathkeeper-proxy.latency":285172,"method":"GET","msg":"completed handling request","remote":"172.18.0.1:53314","request":"/","status":302,"text_status":"Found","time":"2020-03-10T18:33:49Z","took":285172}
    oathkeeper_1                  | [cors] 2020/03/10 18:33:49 Handler: Actual request
    oathkeeper_1                  | [cors] 2020/03/10 18:33:49   Actual request no headers added: missing origin
     ```

2. Accept: application/json

    ```bash
    $ curl -H 'Accept: application/json' http://127.0.0.1:4455/
    {"error":{"code":401,"status":"Unauthorized","message":"Access credentials are invalid"}}
    ```

    ```
    oathkeeper_1                  | [cors] 2020/03/10 18:35:49 Handler: Actual request
    oathkeeper_1                  | [cors] 2020/03/10 18:35:49   Actual request no headers added: missing origin
   oathkeeper_1                  | {"level":"info","method":"GET","msg":"started handling request","remote":"172.18.0.1:53318","request":"/","time":"2020-03-10T18:35:49Z"}
   oathkeeper_1                  | {"error":"Access credentials are invalid","granted":false,"http_host":"127.0.0.1:4455","http_method":"GET","http_url":"http://127.0.0.1:4455/","http_user_agent":"curl/7.64.1","level":"warning","msg":"No authentication handler was responsible for handling the authentication request","reason_id":"authentication_handler_no_match","rule_id":"ory:kratos-selfservice-ui-node:protected","time":"2020-03-10T18:35:49Z"}
    oathkeeper_1                  | {"error":"Access credentials are invalid","granted":false,"http_host":"127.0.0.1:4455","http_method":"GET","http_url":"http://127.0.0.1:4455/","http_user_agent":"curl/7.64.1","level":"warning","msg":"Access request denied","time":"2020-03-10T18:35:49Z"}
    oathkeeper_1                  | {"code":401,"debug":"","details":{},"error":"Access credentials are invalid","level":"error","msg":"An error occurred while handling a request","reason":"","request-id":"","status":401,"time":"2020-03-10T18:35:49Z","trace":"Stack trace: \ngithub.com/ory/oathkeeper/proxy.(*RequestHandler).HandleRequest\n\t/home/circleci/project/proxy/request_handler.go:241\ngithub.com/ory/oathkeeper/proxy.(*Proxy).Director\n\t/home/circleci/project/proxy/proxy.go:134\nnet/http/httputil.(*ReverseProxy).ServeHTTP\n\t/usr/local/go/src/net/http/httputil/reverseproxy.go:206\ngithub.com/urfave/negroni.Wrap.func1\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:46\ngithub.com/urfave/negroni.HandlerFunc.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29\ngithub.com/urfave/negroni.middleware.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38\ngithub.com/ory/x/reqlog.(*Middleware).ServeHTTP\n\t/go/pkg/mod/github.com/ory/x@v0.0.91/reqlog/middleware.go:140\ngithub.com/urfave/negroni.middleware.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38\ngithub.com/ory/x/metricsx.(*Service).ServeHTTP\n\t/go/pkg/mod/github.com/ory/x@v0.0.91/metricsx/middleware.go:261\ngithub.com/urfave/negroni.middleware.ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38\ngithub.com/urfave/negroni.(*Negroni).ServeHTTP\n\t/go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:96\ngithub.com/rs/cors.(*Cors).Handler.func1\n\t/go/pkg/mod/github.com/rs/cors@v1.6.0/cors.go:207\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2007\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2802\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1890\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357","writer":"JSON"}
    oathkeeper_1                  | {"level":"info","measure#oathkeeper-proxy.latency":745776,"method":"GET","msg":"completed handling request","remote":"172.18.0.1:53318","request":"/","status":401,"text_status":"Unauthorized","time":"2020-03-10T18:35:49Z","took":745776}
    ```

The 6th log line in the second example doesn't appear in the first example. Is it possible that there's a bug in the redirect handler that's triggered by JSON requests?

**Comment by amorey** [comment]
I think I figured out what's going on. I'm using Oathkeeper through Ambassador AuthService which doesn't forward the `Accept` HTTP header by default. On the Oathkeeper side, my redirect error handler was only configured for `Accept: text/html` which it wasn't receiving from Ambassador so it used the json error handler to return a response which triggered the stack trace error message in the logs.

Using a redirect handler with the Ambassador AuthService integration seems like a pretty common use case so if you want to prevent another developer from running into this issue, I'd add a comment about configuring AuthService to allow the `Accept` header. Also, the Oathkeeper API documentation says that the `/decisions` endpoint doesn't return 301/302 status codes which led me in the wrong direction when I was debugging.
https://www.ory.sh/docs/oathkeeper/sdk/api#access-control-decision-api

Thanks again for your quick response to my original question.

**Comment by aeneasr** [comment]
> Using a redirect handler with the Ambassador AuthService integration seems like a pretty common use case so if you want to prevent other developers from running into this issue, I'd add a comment about configuring AuthService to allow the Accept header. Also, the Oathkeeper API documentation says that the /decisions endpoint doesn't return 301/302 status codes which led me in the wrong direction when I was debugging:

Makes sense, we can definitely add that in the documentation or even throw a LOG WARN if the accept header is missing?

**Comment by amorey** [comment]
I don't think it's necessary to WARN if the accept header is missing (it could be present but misconfigured) but I think it would be useful to WARN if the request doesn't match a configured error handler. It would also be useful to have the option to see request header details in the DEBUG log.

Thinking back on this issue I think the source of my confusion was that the word "error" is used in three different potential contexts (ERROR level log messages, Oathkeeper error handlers, internal server errors). When I saw the stack trace in the logs I debugged the problem as if it was an internal server error but if I had known it was a runtime issue with the error handler matcher then I think I could have figured it out more quickly.

**Comment by christian-roggia** [comment]
I would like to throw in my two cents regarding an issue I encountered which is strictly related to this one:

> I spent a good portion of my day debugging something related to this issue, I got really confused due to the `authentication_handler_no_match` error. I spent most of the time trying to understand what I misconfigured with my `authenticators` and I was wrongly assuming that oathkeeper was not picking up the `cookie_session` authenticator correctly.
> 
> The error `"No authentication handler was responsible for handling the authentication request","reason_id":"authentication_handler_no_match"` sounds to me like there is no authenticator configured correctly capable of picking up the request, while (from my understanding) the correct interpretation would be that the request did not satisfy any authenticator available (i.e. got denied from all of them). This would be pretty much the same as getting a "**401 - Unauthorized**".

On top of the `authentication_handler_no_match` I was not understanding why the `redirect` handler was not working properly either, assuming something really wrong was going on with my configuration. I am grateful this issue was here and actually provided a solution to the issue. I am too using oathkeeper behind **Ambassador**.

I am thinking about filling some Troubleshooting documentation related to Ambassador / Ambassador Edge Stack (as this is what we use internally) as there have been multiple issues and documentation online is missing for oathkeeper + Ambassador (and the medium articles available are too obsolete for both applications).

**Comment by aeneasr** [comment]
Yes, I also struggled with redirects recently - it's a lot of trial and error and re-reading the code/docs. How could we address the debugability? Is it just the error messages? Or more logs?

**Comment by aeneasr** [comment]
How about we add something that allows us to test access rules? I could imagine `oathkeeper test rules my-rules.json` which you could then make `-curl` requests and would see very detailed trace info in the logs and also some type expected/actual behavior?

**Comment by christian-roggia** [comment]
I would personally expect **a whole lot** more logs when verbosity is set to `debug`. We will soon move to the full integration of oathkeeper with our own authenticators and authorizers, I will follow up with what we believe is missing in the logs as well as general debuggability.

**Comment by aeneasr** [comment]
Yeah I agree - also happy to accept PRs that improve debuggability. We can also use `trace`.

**Comment by xlanor** [comment]
> I would like to throw in my two cents regarding an issue I encountered which is strictly related to this one:
> 
> 
> 
> > I spent a good portion of my day debugging something related to this issue, I got really confused due to the `authentication_handler_no_match` error. I spent most of the time trying to understand what I misconfigured with my `authenticators` and I was wrongly assuming that oathkeeper was not picking up the `cookie_session` authenticator correctly.
> 
> > 
> 
> > The error `"No authentication handler was responsible for handling the authentication request","reason_id":"authentication_handler_no_match"` sounds to me like there is no authenticator configured correctly capable of picking up the request, while (from my understanding) the correct interpretation would be that the request did not satisfy any authenticator available (i.e. got denied from all of them). This would be pretty much the same as getting a "**401 - Unauthorized**".
> 
> 
> 
> On top of the `authentication_handler_no_match` I was not understanding why the `redirect` handler was not working properly either, assuming something really wrong was going on with my configuration. I am grateful this issue was here and actually provided a solution to the issue. I am too using oathkeeper behind **Ambassador**.
> 
> 
> 
> I am thinking about filling some Troubleshooting documentation related to Ambassador / Ambassador Edge Stack (as this is what we use internally) as there have been multiple issues and documentation online is missing for oathkeeper + Ambassador (and the medium articles available are too obsolete for both applications).

@christian-roggia , I was googling for the same error and came across this issue. Likewise, we assumed that this error was coming from cookie_session not being selected as an authenticator, until we took a closer look after seeing this post.

Can I know if you've written the AES documentation for this yet? I have not integrated it behind authservice yet

**Comment by christian-roggia** [comment]
@xlanor we temporarily switched to Istio, if you need an example of Filter / FilterPolicy to setup Ambassador properly I can provide it to you, but I think another ticket/discussion should be opened as this one is not strictly related to Ambassador.
